src/Shared/Components/DynamicDataTable/types.ts

```diff
@@ -0,0 +1,246 @@
+        | 'disableOnBlurResizeToMinHeight'
+        | 'refVar'
+        | 'dependentRef'
+    >
+    [DynamicDataTableRowDataType.DROPDOWN]: Omit<
+        SelectPickerProps<string, false>,
+        'inputId' | 'value' | 'onChange' | 'fullWidth' | 'isDisabled'
+    >
+    [DynamicDataTableRowDataType.SELECT_TEXT]: Omit<
+        SelectTextAreaProps,
+        'value' | 'onChange' | 'inputId' | 'isDisabled' | 'dependentRef' | 'refVar' | 'textAreaProps'
+    > & {
+        textAreaProps?: Omit<
+            SelectTextAreaProps['textAreaProps'],
+            'className' | 'disableOnBlurResizeToMinHeight' | 'minHeight' | 'maxHeight'
+        >
+    }
+    [DynamicDataTableRowDataType.BUTTON]: Pick<
+        DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>,
+        'onClick'
+    > & {
+        icon?: ReactNode
+        text: string
+    }
+}
+
+type DynamicDataTableCellData<T extends keyof DynamicDataTableCellPropsMap = keyof DynamicDataTableCellPropsMap> =
+    T extends keyof DynamicDataTableCellPropsMap ? { type: T; props: DynamicDataTableCellPropsMap[T] } : never
+
+/**
+ * Type representing a key-value row.
+ * @template K - A string representing the key type.
+ */
+export type DynamicDataTableRowType<K extends string, CustomStateType = Record<string, unknown>> = {
+    data: {
+        [key in K]: {
+            value: string
+            disabled?: boolean
+            /** An optional boolean indicating if an asterisk should be shown. */
+            required?: boolean
+        } & DynamicDataTableCellData
+    }
+    id: string | number
+    /** */
+    customState?: CustomStateType
+}
+
+type DynamicDataTableMask<K extends string> = {
+    [key in K]?: boolean
+}
+
+type DynamicDataTableCellIcon<K extends string> = {
+    [key in K]?: (row: DynamicDataTableRowType<K>) => ReactNode
+}
+
+/**
+ * Interface representing the properties for the dynamic data table component.
+ * @template K - A string representing the key type.
+ */
+export type DynamicDataTableProps<K extends string> = {
+    /**
+     * An array containing the headers for the data table. \
+     * Each header defines a column with its label, key, width, and optional settings.
+     */
+    headers: DynamicDataTableHeaderType<K>[]
+    /**
+     * An array of rows where each row contains data corresponding to the table headers.
+     */
+    rows: DynamicDataTableRowType<K>[]
+    /** Optional configuration for sorting the table. */
+    sortingConfig?: {
+        sortBy: K
+        sortOrder: SortingOrder
+        handleSorting: () => void
+    }
+    /** An optional mask to hide the values of the cell. */
+    maskValue?: DynamicDataTableMask<K>
+    /** Optional configuration for displaying an icon in the leading position of a cell. */
+    leadingCellIcon?: DynamicDataTableCellIcon<K>
+    /** Optional configuration for displaying an icon in the trailing position of a cell. */
+    trailingCellIcon?: DynamicDataTableCellIcon<K>
```

Comment:
by naming icon we are telling dev to use only icon here 
discussed with product as well , they are also following this nomenclature

we cannot restrict icon type since that icon can be used to perform actions